include/nbl/asset/utils/CHLSLCompiler.h

```diff
@@ -38,9 +42,15 @@ class NBL_API2 CHLSLCompiler final : public IShaderCompiler
 		//	return "";
 		//}
 
+		std::string preprocessShader(std::string&& code, IShader::E_SHADER_STAGE& stage, const SPreprocessorOptions& preprocessOptions) const override;
+
 	protected:
 
-		void insertExtraDefines(std::string& str, const core::SRange<const char* const>& defines) const override;
+		void insertIntoStart(std::string& code, std::ostringstream&& ins) const override;
+
+		// TODO do we want to use ComPtr? 
```

Comment:
Lets use ComPtr for IDXXXX pointers, its standard practice in dx as The ComPtr uses the interfaces (the I in IDXC) functions (like our smartrefctd and IRefCounted). Also its safer because we may not know what calls happen. And using a unique ptr will call the destructor when it goes out of scope which may do the wrong thing. Use ComPtr for Dx Com objects

Reply:
I think we could use a raw pointer though, but lets be safe.